ring/replication_set.go

```diff
@@ -394,6 +395,80 @@ type instanceResult[T any] struct {
 	instance *InstanceDesc
 }
 
+// DoForReplicationSets runs fn for each ring.ReplicationSet in parallel, and returns a list with all merged results.
+// If any invocation returns an error, DoForReplicationSets returns without waiting for the other invocations to finish.
+func DoForReplicationSets[T any](ctx context.Context, sets []ReplicationSet, fn func(context.Context, ReplicationSet) ([]T, error)) ([]T, error) {
```

Comment:
i realized this does very similar job to `concurrency.ForEachJob`. Should we move `DoForReplicationSets` there and make the interface more similar to `ForEachJob(context.Context, int)`? At which point we can even reuse `ForEachJob` and remove most of the code of this function

Reply:
Very true. I've introduced a new (simpler) `ForEachJobMergeResults()` in https://github.com/grafana/dskit/pull/486/commits/7ebe017a47919e60e73fda0c5d832f1a3f58732e. WDYT?